### ISSUE TITLE
Beginner Question: Can I use this plugin to "transclude" part of my code?

### ISSUE THREAD

**Issue by ReaderGuy42** [saml-dev/obsidian-custom-js#38]
I feel like I'm *almost* getting it, but not quite.
I think I just don't understand the documentation well enough.

I have a file called scriptbits.js in a folder (in my vault) called Script, so I've put that in the plugin settings, and the `const` thing in the code block.

This is what my code looks like:
```dataviewjs
const {ScriptBits} = customJS


dv.paragraph("# I've read  " + (pagesum) + " pages.")
```

And this is what's currently in scriptbits.js:
```
class ScriptBits {
	PageSum1(s) {
			let pagesum = 0;
   for(let i = 0; i < dv.pages('"Book Log/2022"').where(p => p.Medium == "Book" || p.Medium == "eBook").length; i++) {
     if(dv.pages('"Book Log/2022"').where(p => p.Medium == "Book" || p.Medium == "eBook")[i].Length) {
       pagesum += dv.pages('"Book Log/2022"').where(p => p.Medium == "Book" || p.Medium == "eBook")[i].Length;
         }
   }

	}
  }
 ```

But then I still get the error `pagesum not defined`. What am I missing? Is this possible with your plugin?

Thanks! See here for more of what I'm trying to achieve: https://forum.obsidian.md/t/is-it-possible-to-transclude-code-blocks-somehow/50973?u=readerguy42

**Comment by ReaderGuy42** [comment]
Looking at the examples in the Readme, I'm trying something like this now:
```
```dataviewjs
const {ScriptBits} = customJS
let BookCount = (dv.pages('"Book Log/2022"').where(p => ScriptBits.PageMedia).length);

dv.paragraph("# This year, I've read " + BookCount + " books. ")
```
```
and this in the ScriptBit:
```
class ScriptBits {
	PageMedia(s) {
                p.Medium == "Book" || p.Medium == "eBook"
	}
  }
 ```
This gives me **a** number. Writing this I realized that it was just outputting the number of files in that folder... but at least I'm not getting an error currently, though not what I'm going for.

For reference, more than half of my book notes are `Medium: Audiobook` so that's why the total number of notes in that folder doesn't do much.

**Comment by saml-dev** [comment]
You have a couple syntax errors.

you're not calling your function, you're just returning it.
```
.where(p => ScriptBits.PageMedia(p))
```
alternatively, you could just pass the function itself instead of using an arrow function
```
.where(ScriptBits.PageMedia)
```

You're not returning anything from the PageMedia function
```
return p.Medium == "Book" ...
```

**Comment by saml-dev** [comment]
You also don't need the outer parenthesis around dv.pages, and I'm not sure why you have `"'Book Log/2022'"` surrounded by both single and double quotes

**Comment by saml-dev** [comment]
Closing for now as not customJS related

**Comment by ReaderGuy42** [comment]
Sorry for the delay. Even if it's not directly related to this plugin, could I ask another question anyway?

I've tried following what you said, and now have this in my `.js` file:

```
class ScriptBits {
	PageMedia(s) {
               return p.Medium == "Book"
	}
  }
 ```
and this in the dataview:
```
```dataviewjs
const {ScriptBits} = customJS
let BookCount = (dv.pages('"Book Log/2022"').where(ScriptBits.PageMedia).length);

dv.paragraph("# This year, I've read " + BookCount + " books. ")

```

I'm still getting a `ReferenceError: p is not defined at PageMedia`

If you don't have time, maybe you could just point me towards a manual page or something like that?
Thanks!!

**Comment by saml-dev** [comment]
Your `PageMedia` function is taking a parameter called `s` and then you're refering a non-existent variable called `p` when you say `p.Medium`

**Comment by saml-dev** [comment]
All of your errors so far have been javascript syntax errors, so if you want to learn then there are many great resources out there for learning JS. I personally like YouTube as a learning resource.

**Comment by ReaderGuy42** [comment]
I'll look into that, thanks :)

**Comment by ReaderGuy42** [comment]
Now I understand what you meant by not returning anything with my earlier attempts. 

I've now also gotten that it's easier to just `return` the function rather than just an argument, i.e. `dv.pages(\'\"Book Log\/2022\"\').where(p => p.Medium == \"Book\").length`. The `"'single quotes"'` around Book Log are necessary to tell it the folder title. 

However if I escape the quotation marks with `\` or backticks, the text is passed on to dataviewjs but is then not processed.
How can I pass the string on to dataview but still have it process it?

 I hope that this is more relevant to your plugin!!

**Comment by saml-dev** [comment]
This is still just javascript 😁 but it's ok. I'll respond when I can and not when I can't.

You don't need both kinds of quotes. Just use `dv.pages("Book Log 2022")`

**Comment by ReaderGuy42** [comment]
That's fine, this is no rush. Unfortunately that doesn't work on my end. If I leave of the single quotes, I get this error:
```
Evaluation Error: Error: Failed to parse query in 'pagePaths': Error: 
-- PARSING FAILED --------------------------------------------------

> 1 | Book Log/2022
    | ^

Expected one of the following: 

'!', '(', '-', 'csv(', 'outgoing(', file link, string, tag ('#hello/stuff')
```
That is in my control group btw, which looks like this when it works:
`dv.paragraph("# In 2022 I read " + dv.pages('"Book Log/2022"').where(p => p.Medium == "Book").length + " books. ")`

My Custom-js code still also just outputs the intended code without processing it. 

no rush!! thanks :)

**Comment by saml-dev** [comment]
Hm it seems your use case is to list books you read in 2022. Instead of trying to use `dataviewjs` have you tried just using regular `dataview` blocks? It will be simpler

**Comment by ReaderGuy42** [comment]
What I'm trying to do here is to get the number of books, which I wasn't able to achieve with regular `dataview`. Someone (can't remember who) recommended doing it with `dataviewjs`.

This gets more complicated because I also separate my book notes by medium so I can get a "total pages read" for mediums like physical books, ebooks, graphic novels, academic articles, etc.; and a "total hours listened" for audiobooks and podcasts.

I can achieve this by hardcoding the mediums into the `dataviewjs` query, e.g. `dv.paragraph("# In 2022 I read " + dv.pages('"Book Log/2022"').where(p => p.Medium == "Book").length + " books. ")` or creating a variable called `BookCount` so that it looks like `dv.paragraph("# This year, I've read " + BookCount + " books. ")`

But I was hoping to be able to transclude this. The reason being that if I add a medium, e.g. I just added `GraphicNovel` to the mix, I then have to go through and edit all the individual instances of where I hardcoded the Mediums in.

So my goal is to have one `.js` file that looks like
```
class ScriptBits {
	PageMedia() {
              return `dv.pages('"Book Log/2022"').where(p => p.Medium == "Book" ||  p.Medium == "GraphicNovel").length`
	}
  }
 ```
and then in the queries I just refer to that function so I don't have to edit everything again. Because this has become a little unwieldy if I try embedding that into e.g. my total sum of pages:

```
let pagesum = 0;
   for(let i = 0; i < dv.pages('"Book Log/2022"').where(p => p.Medium == "Book" || p.Medium == "eBook" || p.Medium == "Article" || p.Medium == "Chapter" || p.Medium == "Monograph" || p.Medium == "GraphicNovel").length; i++) {
     if(dv.pages('"Book Log/2022"').where(p => p.Medium == "Book" || p.Medium == "eBook" || p.Medium == "Article" || p.Medium == "Chapter" || p.Medium == "Monograph" || p.Medium == "GraphicNovel")[i].Length) {
       pagesum += dv.pages('"Book Log/2022"').where(p => p.Medium == "Book" || p.Medium == "eBook" || p.Medium == "Article" || p.Medium == "Chapter" || p.Medium == "Monograph" || p.Medium == "GraphicNovel")[i].Length;
         }
   }
```
Hope that clears up my usecase a bit! Thanks so much for your help :)
